pkg/generate/dockercompose/generate.go

```diff
@@ -0,0 +1,537 @@
+				joins[k].Insert(target)
+				volumesFrom[k] = append(volumesFrom[k], target)
+			case 3:
+				joins[k].Insert(parts[1])
+				volumesFrom[k] = append(volumesFrom[k], parts[1])
+			}
+		}
+	}
+	joinOrder := sets.NewString()
+	for k := range joins {
+		joinOrder.Insert(k)
+	}
+	var colocated []sets.String
+	for _, k := range joinOrder.List() {
+		set := joins[k]
+		matched := -1
+		for i, existing := range colocated {
+			if set.Intersection(existing).Len() == 0 {
+				continue
+			}
+			if matched != -1 {
+				return nil, fmt.Errorf("%q belongs with %v, but %v also contains some overlapping elements", k, set, colocated[matched])
+			}
+			existing.Insert(set.List()...)
+			matched = i
+			continue
+		}
+		if matched == -1 {
+			colocated = append(colocated, set)
+		}
+	}
+
+	// identify service aliases
+	aliases := make(map[string]sets.String)
+	for _, v := range p.Configs {
+		for _, s := range v.Links.Slice() {
+			parts := strings.SplitN(s, ":", 2)
+			if len(parts) != 2 || parts[0] == parts[1] {
+				continue
+			}
+			set := aliases[parts[0]]
+			if set == nil {
+				set = sets.NewString()
+				aliases[parts[0]] = set
+			}
+			set.Insert(parts[1])
+		}
+	}
+
+	// find and define build pipelines
+	for _, k := range serviceOrder.List() {
+		v := p.Configs[k]
+		if len(v.Build) == 0 {
+			continue
+		}
+		if _, ok := builds[v.Build]; ok {
+			continue
+		}
+		var base, relative string
+		for _, s := range bases {
+			if !strings.HasPrefix(v.Build, s) {
+				continue
+			}
+			base = s
+			path, err := filepath.Rel(base, v.Build)
+			if err != nil {
+				return nil, fmt.Errorf("path is not relative to base: %v", err)
+			}
+			relative = path
+			break
+		}
+		if len(base) == 0 {
+			return nil, fmt.Errorf("build path outside of the compose file: %s", v.Build)
+		}
+
+		// if this is a Git repository, make the path relative
+		if root, err := git.NewRepository().GetRootDir(base); err == nil {
+			if relative, err = filepath.Rel(root, filepath.Join(base, relative)); err != nil {
+				return nil, fmt.Errorf("unable to find relative path for Git repository: %v", err)
+			}
+			base = root
+		}
+		buildPath := filepath.Join(base, relative)
+
+		// TODO: what if there is no origin for this repo?
+
+		glog.V(4).Infof("compose service: %#v", v)
+		repo, err := app.NewSourceRepositoryWithDockerfile(buildPath, "")
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		repo.BuildWithDocker()
+
+		info := repo.Info()
+		if info == nil || info.Dockerfile == nil {
+			errs = append(errs, fmt.Errorf("unable to locate a Dockerfile in %s", v.Build))
+			continue
+		}
+		node := info.Dockerfile.AST()
+		baseImage := dockerfileutil.LastBaseImage(node)
+		if len(baseImage) == 0 {
+			errs = append(errs, fmt.Errorf("the Dockerfile in the repository %q has no FROM instruction", info.Path))
+			continue
+		}
+
+		var ports []string
+		for _, s := range v.Ports {
+			container, _ := extractFirstPorts(s)
+			ports = append(ports, container)
+		}
```

Comment:
I don't know.  I think we want to let some define EXACTLY what they want to
expose.  That way if you expose everything in the dockerfile, but layer on
top of the dockerfile, you can still control.  So if no ports are set,
check the image.

On Fri, Apr 29, 2016 at 1:39 PM, Cesar Wong <notifications@github.com>
wrote:

> In pkg/generate/dockercompose/generate.go
> <https://github.com/openshift/origin/pull/7690#discussion_r61617130>:
>
> > +		if info == nil || info.Dockerfile == nil {
> > +			errs = append(errs, fmt.Errorf("unable to locate a Dockerfile in %s", v.Build))
> > +			continue
> > +		}
> > +		node := info.Dockerfile.AST()
> > +		baseImage := dockerfileutil.LastBaseImage(node)
> > +		if len(baseImage) == 0 {
> > +			errs = append(errs, fmt.Errorf("the Dockerfile in the repository %q has no FROM instruction", info.Path))
> > +			continue
> > +		}
> > +
> > +		var ports []string
> > +		for _, s := range v.Ports {
> > +			container, _ := extractFirstPorts(s)
> > +			ports = append(ports, container)
> > +		}
>
> Should these be merged with exposed ports in the Compose.yaml?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7690/files/b85cc800bdcb1e7db8f8c37dd3e85c21b5023f0a#r61617130>
>